### PR TITLE
e_os.h removal from other headers

### PR DESCRIPTION
**Pull request by paulidale**
Removed e_os.h from all bar two headers (apps/apps.h and crypto/bio/bio_lcl.h).
Added e_os.h into the files that need it now.
Directly reference internal/nelem.h when required.

- [x] tests are added or updated

Refer: #4174

### CODE REVIEW

**Comment by paulidale** [comment]
AppVeyor failure is relevant.

**Comment by richsalz** [comment]
travis failure, https://ci.appveyor.com/project/openssl/openssl/build/master.12576/job/fx3sk3rnryrelogy#L1968, looks relevant.

**Comment by dot-asm** [comment]
There is certain ambiguity in this. On Windows timeval is defined in winsock[2].h but API kind of suggests that application doesn't have to include any of "networking" headers. Indeed, whole idea behind BIO_sock things is to hide system specifics, or in other words provide system-neutral interface. And trouble is that we can't include winsock[2].h in our public headers, because it's position-sensitive. In sense that it has to be included in certain order, specifically in relation to windows.h, and we don't want to mess with application code. It's simpler/more appropriate to claim that it's application's responsibility to ensure wisock[2].h is included. Correct solution to the problem is to define own time interval time, or rather own time type (it was suggested that it should have sub-second granularity) that would work as both absolute and relative for all interfaces that take time. Unfortunately this is not something we can do in minor release. For this reason, if we aim for removing reliance on e_os.h in apps/\* (which is good thing) we have to explicitly include winsock[2].h...

**Comment by paulidale** [comment]
I've put the `e_os.h` include back into `ssl_locl.h` which fixes the builds.

The struct timeval issue will need a solution in due course (or be accepted as is).

**Comment by dot-asm** [comment]
If we consider e_os.h for example in ec_key.c, we'll see that it depends on REF_PRINT_COUNT and REF_ASSERT_ISNT definitions. None of which are actually OS-specific. Or in other words both are generic C constructs. One can even wonder if it's appropriate to move these two to include/internal/refcount.h or even cryptlib.h? But for the moment this is not actually a suggestion, rather a question! However, if it's answered, it would be possible to omit even more e_os.h includes, quite a bit actually. So that question is definitely worth answering, One can argue that it can be done in separate request, or do it right away, In latter case moving to cryptlib.h would probably be least invasive...

**Comment by richsalz** [comment]
yes, that would be interesting.  BTW GETPID_IS_MEANINGLESS can go away

**Comment by paulidale** [comment]
Most of the source files that need e_os.h only require one piece of functionality from it.  REF_PRINT_COUNT is one of the commoner ones in used this way.  Should the pieces go into their own headers or cryptlib.h?  It seems unwise to copy all of the pieces over to cryptlib.h, lest it clutters the same way e_os.h is.

We're down to one use of OSSLzu in the sources, another candidate for removal.

**Comment by richsalz** [comment]
IMO: don't add to cryptlib, create one or more new files.\
And BTW, the define for getpid can also go away.

**Comment by richsalz** [comment]
And X_OK can go away :)

**Comment by levitte** [comment]
I'm noticing a style inconsistency, where you include `internal/refcount.h` with quotes and `internal/nelem.h` with angle brackets (and change one inclusion of `e_os.h` from quoted to bracketed?).  Please stay with one style.  Personally, I prefer quotes for internal headers.

**Comment by paulidale** [comment]
#include <internal/nelem.h> changed to #include "internal/nelem.h"

File: ssl/ssl_locl.h [hunk 1]
```diff
@@ -16,7 +16,7 @@
 # include <string.h>
 # include <errno.h>
 
-# include "e_os.h"
+# include <e_os.h>              /* struct timeval for Windows */
```
Comment (levitte) [inline ssl/ssl_locl.h]: Can I convince you to put back the quotes? ;-)

Comment (paulidale) [inline ssl/ssl_locl.h]: Yes.  I don't know why I changed it.

**Comment by paulidale** [comment]
Fixed and rebased to avoid merge problems.

**Comment by richsalz** [comment]
build failures related?  E.g., see https://travis-ci.org/openssl/openssl/jobs/267370657#L1322

**Comment by paulidale** [comment]
Fix build issue.

File: crypto/asn1/ameth_lib.c [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 #include <stdio.h>
 #include "internal/cryptlib.h"
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/asn1/ameth_lib.c]: This one is redundant.

File: crypto/conf/conf_mod.c [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include <ctype.h>
 #include <openssl/crypto.h>
 #include "internal/cryptlib.h"
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/conf/conf_mod.c]: On *side note* this one seems to be dependency on OPENSSL_CONF which isn't OS-specific.

File: crypto/dso/dso_win32.c [hunk 4]
```diff
@@ -8,6 +8,7 @@
  */
 
 #include "dso_locl.h"
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/dso/dso_win32.c]: On *side note* there is ambiguity here. Remember I mentioned that winsock[2].h is position-sensitive? And since e_os.h includes it, it, e_os.h, becomes position-sensitive too. In more practical terms in order to place e_os.h in specific position, you have to be sure that none of earlier headers include <windows.h>. Fool-proof way is to always include e_os.h first. Yes, you can observe multiple cases when e_os.h is not first, but this only means that it happens to work out in that particular cases. Besides, "position sensitivity" is rather about what gets declared, not compilation failures, and in those cases where e_os.h is not first (or early enough) you're off the hook if you don't need those missing declarations. But arguably one can wonder if it still makes sense to formulate some kind of rule, such as "e_os.h always first", anyway. You know, because it's easier to apply simple rule, than to figure out if position is right on per-case basis... Once again, this was a side note...

File: crypto/ec/ec_key.c [hunk 5]
```diff
@@ -11,6 +11,7 @@
 #include "internal/cryptlib.h"
 #include <string.h>
```
Comment (dot-asm) [inline crypto/ec/ec_key.c]: In the context of previous side note. I have some wage memory that it, header ordering rule, was indirectly discussed at some point. It was formulated as "cryptlib.h over everything else", and we do see instance of it here. Do note that cryptlib.h did include e_os.h till now...

File: ssl/statem/statem.c [hunk 6]
```diff
@@ -8,6 +8,7 @@
  */
 
 #include <openssl/rand.h>
+#include "e_os.h"
```
Comment (dot-asm) [inline ssl/statem/statem.c]: This one also seems to be dependency on (not-really-os-specific) OPENSSL_CONF...

File: crypto/x509v3/v3_ncons.c [hunk 7]
```diff
@@ -9,6 +9,7 @@
 
 #include <stdio.h>
 #include "internal/cryptlib.h"
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: This one is OPENSSL_CONF too, surely there are few more...

File: crypto/dso/dso_win32.c (hunk 4 above)
Comment (dot-asm) [inline crypto/dso/dso_win32.c]: But on less side-note-ish note, it wouldn't be inappropriate to aim for order *preservation* in this request. And since original cryptlib.h was including e_os.h *prior* all other openssl/\*.h it would be more appropriate to keep this pattern. I mean in multiple cases in this request we observe
```
#include "internal/cryptlib.h"
#include "e_os.h"
```
while less intrusive order would be
```
#include "e_os.h"
#include "internal/cryptlib.h"
```
At least we could be more confident that there are no new subtle issues...

File: crypto/conf/conf_mod.c (hunk 3 above)
Comment (paulidale) [inline crypto/conf/conf_mod.c]: Yes, there are two uses of OPENSSL_CONF in the code base.  Here and `apps/openssl.c`.
I was hoping to unify them rather than keeping them in a header.

If it is to move to a different header, which? `cryptlib.h`?

File: crypto/ec/ec_key.c (hunk 5 above)
Comment (paulidale) [inline crypto/ec/ec_key.c]: Okay, I wasn't aware of this.

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (paulidale) [inline crypto/x509v3/v3_ncons.c]: ?? I don't see it.

File: crypto/asn1/ameth_lib.c (hunk 2 above)
Comment (paulidale) [inline crypto/asn1/ameth_lib.c]: Removed.

File: crypto/dso/dso_win32.c (hunk 4 above)
Comment (paulidale) [inline crypto/dso/dso_win32.c]: I've moved e_os.h to be the very first include, before even the system includes.

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (paulidale) [inline crypto/x509v3/v3_ncons.c]: e_os.h isn't required now and has been removed.

**Comment by paulidale** [comment]
Addressed the review comments.  Thanks.

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (paulidale) [inline crypto/x509v3/v3_ncons.c]: it is required on Windows for `strncasecmp`.  So not removed after all.

File: crypto/asn1/ameth_lib.c (hunk 2 above)
Comment (paulidale) [inline crypto/asn1/ameth_lib.c]: Not removed.  Needed on Windows for strncasecmp.
I've added a comment to indicate this.

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: Ah! I'm not at my desk right now, and am not in position to make adequately weighed assessments. Apologies for distraction.

File: crypto/blake2/blake2b.c [hunk 8]
```diff
@@ -14,10 +14,10 @@
  * can be found at https://blake2.net.
  */
 
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/blake2/blake2b.c]: Even though it might turn to be another diversion, do blake2[bs].c really need e_os.h? I seem to have wage memory that in previous version it was removing and me thinking "oh! that's good, algorithm implementations are last ones that would be OS-specific..."

File: crypto/des/cfb64ede.c [hunk 9]
```diff
@@ -7,8 +7,8 @@
  * https://www.openssl.org/source/license.html
  */
 
-#include "des_locl.h"
 #include "e_os.h"
+#include "des_locl.h"
```
Comment (dot-asm) [inline crypto/des/cfb64ede.c]: One wonder same question [as with blake2] here. This is an algorithm implementation that can be done in pure C, there shouldn't be OS dependencies.

One can probably argue that this, removal of redundant e_os.h inclusions, should be matter for separate request...

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (richsalz) [inline crypto/x509v3/v3_ncons.c]: Does it make sense to create a new header file for all string,strings,memcmp kinds of things?  I'm thinking pull the malloc stuff out of crypto.h, the ctype, etc., stuff out of e_os.h and so on.  It would be a separate PR, but just a thought.

Comment (levitte) [inline crypto/x509v3/v3_ncons.c]: I agree.  `internal/string.h`, anyone?

Comment (dot-asm) [inline crypto/x509v3/v3_ncons.c]: As for separate header. It's fine line. In sense that we don't really want too many of them. At least we don't want as many as standard ones, do we? And considering this example it might be tricky. Note that it started with strncasecmp, but ended with suggestion to create internal/string.h. See the contradiction? strncasecmp is defined strings.h, not string.h... I would like to suggest to proceed with this request without creating new files, collect our thoughts and do something about it in separate request. [To tell a little secret I was going to discuss this already at first meeting, but agenda was jammed.]

Comment (richsalz) [inline crypto/x509v3/v3_ncons.c]: I have a separate PR coming up for internal/string.h ...

File: crypto/blake2/blake2b.c (hunk 8 above)
Comment (paulidale) [inline crypto/blake2/blake2b.c]: I thought I tried these without but don't remember the outcome.  Trying again via a commit to see if the AppVeyor builds break.

File: crypto/des/cfb64ede.c (hunk 9 above)
Comment (paulidale) [inline crypto/des/cfb64ede.c]: I've removed redundant inclusions here already, so they do fit this request.

**Comment by paulidale** [comment]
Those two didn't need e_os.h after all.  Gone now and squashed into earlier commits.

File: crypto/x509v3/v3_ncons.c (hunk 7 above)
Comment (paulidale) [inline crypto/x509v3/v3_ncons.c]: A new PR is good.  There are several instances where e_os.h is included solely for the string functions.  I added some comments to this effect but I'm pretty sure I didn't do all of them.

Breaking e_os.h into a manageable number of other internal include files makes sense to me.  Each inclusion would be more specific about what it requires and it solves the whole monolithic problem this began with.  The key is manageable, it would easily get out of hand.  My `nelem.h` is probably an example of this already :(

File: crypto/blake2/blake2s.c [hunk 10]
```diff
@@ -14,10 +14,10 @@
  * can be found at https://blake2.net.
  */
 
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/blake2/blake2s.c]: If it's not needed in blake2b, there is no way it's needed in blake2s.

File: crypto/x509/by_dir.c [hunk 11]
```diff
@@ -7,12 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+#include "internal/cryptlib.h"
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/x509/by_dir.c]: Flip these two. I mean "e_os.h" should be first.

File: crypto/x509v3/v3_tlsf.c [hunk 12]
```diff
@@ -1,14 +1,15 @@
...
  *
  * Licensed under the OpenSSL license (the "License").  You may not use
  * this file except in compliance with the License.  You can obtain a copy
  * in the file LICENSE in the source distribution or at
  * https://www.openssl.org/source/license.html
  */
 
-#include <stdio.h>
 #include "internal/cryptlib.h"
+#include "e_os.h"
```
Comment (dot-asm) [inline crypto/x509v3/v3_tlsf.c]: Flip these two. Trouble is that internal/cryptlib.h includes a whole bunch of headers and we can't be sure that none of them ends up triggering "position sensitivity" on Windows. Well, one can argue that it's OK since it works(\*), but it's not inappropriate to follow same pattern everywhere.

(\*) after all "position sensitivity" affects which declarations you "see" and if you don't rely on what you don't see, then it's not a problem;

File: ssl/s3_lib.c [hunk 13]
```diff
@@ -9,8 +9,10 @@
  * https://www.openssl.org/source/license.html
  */
 
+#include "e_os.h"
```
Comment (dot-asm) [inline ssl/s3_lib.c]: On side note I wonder what's in s3_lib.c that is [that] OS-specific...

File: ssl/ssl_locl.h [hunk 14]
```diff
@@ -11,12 +11,12 @@
...
+# include "e_os.h"              /* struct timeval for Windows */
 # include <stdlib.h>
 # include <time.h>
 # include <string.h>
 # include <errno.h>
 
-# include "e_os.h"
 # if defined(__unix) || defined(__unix__)
 #  include <sys/time.h>         /* struct timeval for DTLS */
 # endif
```
Comment (dot-asm) [inline ssl/ssl_locl.h]: On side note (and this is more like a mental note for myself) one can wonder if it's appropriate to move this to e_os.h. I mean if Windows has to get timeval from e_os.h, then it would be "prettier" if Unix did too...

File: crypto/blake2/blake2s.c (hunk 10 above)
Comment (paulidale) [reply crypto/blake2/blake2s.c]: Okay, I'll check this out.

File: crypto/x509/by_dir.c (hunk 11 above)
Comment (paulidale) [reply crypto/x509/by_dir.c]: Yes, e_os.h should be first.  Will fix.

File: ssl/s3_lib.c (hunk 13 above)
Comment (paulidale) [reply ssl/s3_lib.c]: I'll check this one too.

File: ssl/ssl_locl.h (hunk 14 above)
Comment (paulidale) [reply ssl/ssl_locl.h]: Good suggestion.  Unless someone objects, I'll look into doing this too.

Comment (dot-asm) [reply ssl/ssl_locl.h]: I'd suggest to do it in separate request. At least they moved socket stuff to separate file, so it might happen that it would be appropriate to move some place else. Myself I need to re-evaluate things with this new header, so I have no suggestions at this point.

**Comment by dot-asm** [comment]
Do note that there are conflicts now, so you have to re-base this.

File: ssl/ssl_locl.h (hunk 14 above)
Comment (paulidale) [reply ssl/ssl_locl.h]: Okay, I'll leave this one for later.

**Comment by paulidale** [comment]
Updated and rebased.

File: ssl/ssl_locl.h (hunk 14 above)
Comment (paulidale) [reply ssl/ssl_locl.h]: I created #4271 so this doesn't entirely disappear.

**Comment by dot-asm** [comment]
Do I wait for check for e_os.h omission in blake2s.c (should work 100% since blake2b.c did) and s3_lib.c (not 100% sure about it)? Or do you expect it to be approved as it is now (and test omission in another request)?

**Comment by paulidale** [comment]
My mistake, I made those changes and then lost them during the squash :(

**Comment by paulidale** [comment]
Fixed now.

**Comment by paulidale** [comment]
Thanks, I'll merge tomorrow (12-13 hours) if there are no conflicts.

**Comment by paulidale** [comment]
Merged.
